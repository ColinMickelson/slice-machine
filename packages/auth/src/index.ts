export { createPrismicAuth } from "./createPrismicAuth";

export { createPrismicAuthMiddleware } from "./createPrismicAuthMiddleware";

export type { PrismicAuth } from "./PrismicAuth";

export type { PrismicUserProfile } from "./types";

export {
	UnauthenticatedError,
	isUnauthenticatedError,
	UnexpectedDataError,
	isUnexpectedDataError,
} from "./errors";
