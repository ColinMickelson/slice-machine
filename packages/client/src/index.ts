import axios, { AxiosError, AxiosPromise, AxiosRequestHeaders } from "axios";
import * as t from "io-ts";

import { UserProfile } from "@slicemachine/core/build/models";
import { SharedSlice } from "@prismicio/types-internal/lib/customtypes/widgets/slices";
import { CustomType } from "@prismicio/types-internal/lib/customtypes/CustomType";

import type { ClientError, UploadParameters } from "./models";
import {
  ApplicationMode,
  getStatus,
  getMessage,
  ApisEndpoints,
  ProductionApisEndpoints,
  StageApisEndpoints,
  AclCreateResult,
} from "./models";

import { getAndValidateResponse } from "./utils";
import { upload } from "./utils/upload";

// exporting models to be used with the Client.
export * from "./models";

// exporting utils so the client calls can be replicated for speficic calls to a consumer.
export * from "./utils";

// exporting the client itself.
export class Client {
  apisEndpoints: ApisEndpoints;
  repository: string | null;
  authenticationToken: string;

  constructor(
    applicationMode: ApplicationMode,
    repository: string | null,
    authenticationToken: string
  ) {
    this.repository = repository;
    this.authenticationToken = authenticationToken;

    if (applicationMode === ApplicationMode.PROD)
      this.apisEndpoints = ProductionApisEndpoints;
    else if (applicationMode === ApplicationMode.STAGE)
      this.apisEndpoints = StageApisEndpoints;
    else {
      // Dev
      this.apisEndpoints = ProductionApisEndpoints;
    }
  }

  // setters to provide flexibility
  updateAuthenticationToken(newToken: string): Client {
    this.authenticationToken = newToken;
    return this
  }

  updateRepository(repository: string | null): Client {
    this.repository = repository;
    return this
  }

  // private methods
  _fetch(
    method: "get" | "post",
    url: string,
    data?: Record<string, unknown>,
    headers?: AxiosRequestHeaders
  ): AxiosPromise {
    return axios({
      method,
      url,
      data,
      headers: {
        Authorization: `Bearer ${this.authenticationToken}`,
        "User-Agent": "slice-machine",
        ...(this.repository ? { repository: this.repository } : {}),
        ...headers,
      },
    });
  }

  _get(url: string): AxiosPromise {
    return this._fetch("get", url);
  }

  _post(url: string, data: Record<string, unknown>): AxiosPromise {
    return this._fetch("post", url, data).catch((error: Error | AxiosError) => {
      const status: number = getStatus(error);
      const message: string = getMessage(error);

      // Making sure the error is typed
      const clientError: ClientError = { status, message };
      return Promise.reject(clientError);
    });
  }

  async validateAuthenticationToken(): Promise<boolean> {
    return this._get(
      `${this.apisEndpoints.Authentication}validate?token=${this.authenticationToken}`
    )
      .then(() => true)
      .catch((error: Error | AxiosError) => {
        const status: number = getStatus(error);
        if (status == 401) return false;

        const message: string = getMessage(error);

        // Making sure the error is typed
        const clientError: ClientError = { status, message };
        return Promise.reject(clientError);
      });
  }

  async refreshAuthenticationToken(): Promise<string> {
    return getAndValidateResponse<string>(
      this._get(
        `${this.apisEndpoints.Authentication}refreshtoken?token=${this.authenticationToken}`
      ),
      "refreshed authentication token",
      t.string
    ).then((newToken) => {
      this.updateAuthenticationToken(newToken);
      return newToken;
    });
  }

  async profile(): Promise<UserProfile> {
    return getAndValidateResponse<UserProfile>(
      this._get(`${this.apisEndpoints.Users}profile`),
      "user profile",
      UserProfile
    );
  }

  async getSlices(): Promise<SharedSlice[]> {
    return getAndValidateResponse<SharedSlice[]>(
      this._get(`${this.apisEndpoints.Models}slices`),
      "shared slices",
      t.array(SharedSlice)
    );
  }

  async getCustomTypes(): Promise<CustomType[]> {
    return getAndValidateResponse<CustomType[]>(
      this._get(`${this.apisEndpoints.Models}customtypes`),
      "custom types",
      t.array(CustomType)
    );
  }

  async insertCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/insert`,
      customType
    );
  }

  async updateCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/update`,
      customType
    );
  }

  async insertSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/insert`, sharedSlice);
  }

  async updateSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/update`, sharedSlice);
  }

  async createImagesAcl(): Promise<AclCreateResult> {
    return getAndValidateResponse<AclCreateResult>(
      this._get(`${this.apisEndpoints.AclProvider}create`),
      "acl",
      AclCreateResult
    );
  }

  async deleteImagesFolderAcl(sliceName: string): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.AclProvider}delete-folder`, {
      sliceName,
    });
  }

  async uploadImageAcl(params: UploadParameters): Promise<number | undefined> {
    return upload(params);
  }
}
