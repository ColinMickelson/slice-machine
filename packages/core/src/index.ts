export * as FileSystem from "./filesystem";
export * as Utils from "./utils";
export * from "./core";
export * as Models from "./models";
export * as Libraries from "./libraries";
export * as Mocks from "./libraries/mocks";
export * from "./auth";
