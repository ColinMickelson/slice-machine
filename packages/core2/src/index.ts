export {
	SliceMachineManager,
	createSliceMachineManager,
} from "./createSliceMachineManager";

export {
	CreateSliceMachineManagerMiddlewareArgs,
	createSliceMachineManagerMiddleware,
} from "./createSliceMachineManagerMiddleware";

export {
	PrismicAuthManager,
	PrismicUserProfile,
	createPrismicAuthManager,
} from "./createPrismicAuthManager";

export {
	CreatePrismicAuthManagerMiddlewareArgs,
	PrismicAuthCheckStatusResponse,
	createPrismicAuthManagerMiddleware,
} from "./createPrismicAuthManagerMiddleware";
