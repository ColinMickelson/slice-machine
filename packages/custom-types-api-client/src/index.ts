export { createCustomTypesAPIClient } from "./createCustomTypesAPIClient";

export type { CustomTypesAPIClient } from "./CustomTypesAPIClient";

export {
	InvalidPayloadError,
	ForbiddenError,
	ConflictError,
	NotFoundError,
	PrismicError,
} from "@prismicio/custom-types-client";
