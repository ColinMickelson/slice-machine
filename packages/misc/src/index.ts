// TODO: Figure out where libs should be
export { decode } from "./decode";
export type { DecodeReturnType } from "./decode";
export { DecodeError } from "./DecodeError";
export {
	readPrismicrc,
	writePrismicrc,
	updatePrismicrc,
	Prismicrc,
} from "./prismicrc";
export {
	getAllStableSliceMachineVersions,
	getReleaseNotesForVersion,
} from "./versions";

// Errors
export { SliceMachineError, isSliceMachineError } from "./SliceMachineError";
export { InternalError, isInternalError } from "./InternalError";

// Constants
export { ApplicationMode } from "./ApplicationMode";
export { APIEndpoints } from "./APIEndpoints";
export { APITokens } from "./APITokens";
