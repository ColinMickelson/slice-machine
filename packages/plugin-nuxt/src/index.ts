export {
  slice,
  story,
  syntax,
  snippets,
  index,
} from "@slicemachine/plugin-vue";

export const framework = "nuxt";
