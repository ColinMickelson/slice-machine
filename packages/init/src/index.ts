#!/usr/bin/env node

import { Utils, FileSystem } from "@slicemachine/core";
import { Tracker } from "./utils/tracker";
import {
  installRequiredDependencies,
  validatePkg,
  maybeExistingRepo,
  createRepository,
  loginOrBypass,
  configureProject,
  displayFinalMessage,
  detectFramework,
  installLib,
} from "./steps";
import { findArgument } from "./utils";

async function init() {
  const cwd = findArgument(process.argv, "cwd") || process.cwd();
  const base = findArgument(process.argv, "base") || Utils.CONSTS.DEFAULT_BASE;
  const lib: string | undefined = findArgument(process.argv, "library");
  const branch: string | undefined = findArgument(process.argv, "branch");
  const tracking =
    findArgument(process.argv, "tracking") === "false" ? false : true;

  console.log(
    Utils.purple(
      "You're about to configure Slicemachine... Press ctrl + C to cancel"
    )
  );

  // verify package.json file exist
  validatePkg(cwd);

  // login
  const user = await loginOrBypass(base);
  if (!user) throw new Error("The user should be logged in!");

  // retrieve tokens for api calls
  const config = FileSystem.PrismicSharedConfigManager.get();

  // detect the framework used by the project
  const frameworkResult = await detectFramework(cwd);

  // select the repository used with the project.
  const { existing, name } = await maybeExistingRepo(
    config.cookies,
    cwd,
    config.base
  );

  if (!existing) {
    await createRepository(name, frameworkResult.value, config);
  }

  const tracker = Tracker.build(
    "JfTfmHaATChc4xueS7RcCBsixI71dJIJ",
    name,
    {
      userId: user.userId,
    },
    tracking
  );

  // install the required dependencies in the project.
  await installRequiredDependencies(cwd, frameworkResult.value);

  const sliceLibPath = lib
    ? await installLib(tracker, cwd, lib, branch)
    : undefined;

  // configure the SM.json file and the json package file of the project..
  configureProject(cwd, base, name, frameworkResult, sliceLibPath, tracking);

  // ask the user to run slice-machine.
  displayFinalMessage(cwd);
}

try {
  void init();
} catch (error) {
  if (error instanceof Error) Utils.writeError(error.message);
  else console.error(error);
}
