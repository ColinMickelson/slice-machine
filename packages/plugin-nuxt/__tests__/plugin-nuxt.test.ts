import { expect, test } from "@jest/globals";
import * as plugin from "../src";
import type { Variations } from "@slicemachine/plugin-middleware";
import { FieldType } from "@slicemachine/plugin-middleware";

describe("@slicemachine/plugin-nuxt", () => {
  test("#framework", () => {
    expect(plugin.framework).toEqual("nuxt");
  });

  test("#syntax", () => {
    expect(plugin.syntax).toEqual("html");
  });

  test("#slice", () => {
    const result = plugin.slice("Foo");
    expect(result.data).toContain('name: "Foo"');
    expect(result).toMatchSnapshot();
  });

  test("#story", () => {
    const variations = [
      {
        id: "default-slice",
        name: "Default slice",
        docURL: "...",
        version: "sktwi1xtmkfgx8626",
        description: "FooSlice",
        primary: {
          title: {
            type: "StructuredText" as const,
            config: {
              single: "heading1",
              label: "Title",
              placeholder: "This is where it all begins...",
            },
          },
          description: {
            type: "StructuredText",
            config: {
              single: "paragraph",
              label: "Description",
              placeholder: "A nice description of your feature",
            },
          },
          "": {
            config: {
              label: "",
              constraint: {},
              thumbnails: [],
            },
            type: "Image",
          },
        },
        items: {},
      },
    ] as Variations;
    const result = plugin.story("../slices/foo", "FooBar", variations);
    expect(result.filename).toEqual("index.stories.vue");
    expect(result.data).toContain("import MyComponent from '../slices/foo'");

    expect(result.data).toContain("title: 'FooBar'");

    expect(result.data).toContain(`export const default-slice = () => ({`);

    expect(result.data).toContain(
      `default-slice.storyName = '${variations[0].name}'`
    );

    expect(result).toMatchSnapshot();
  });

  describe("#index", () => {
    test("it creates an index.js file", () => {
      const result = plugin.index(["Foo", "Bar", "Baz"]);
      expect(result.filename).toEqual("index.js");
      expect(result.data).toContain("export { default as Foo } from './Foo'");
      expect(result.data).toMatchSnapshot();
    });
  });

  describe("#snippets", () => {
    const SupportedFields = Object.values(FieldType).filter(
      (type) => type !== FieldType.Group && type !== FieldType.IntegrationFields
    );

    const suported = SupportedFields.map((field, i) => [
      field,
      `slices[${i}].data`,
    ]);

    test.each(suported)(
      "supported type %s should be truthy",
      (type, fieldText) => {
        const result = plugin.snippets({ type, fieldText });
        expect(result).toBeTruthy();
        expect(result).toMatchSnapshot();
      }
    );

    test.each(suported.map((d) => [...d, true, false]))(
      "snippet %s with useKey",
      (type, fieldText, useKey, isRepeatable) => {
        const result = plugin.snippets({
          type,
          fieldText,
          useKey,
          isRepeatable,
        });
        expect(result).toBeTruthy();
        expect(result).toMatchSnapshot();
      }
    );

    const UnsupprtedFields = Object.values(FieldType).filter(
      (type) => SupportedFields.includes(type) === false
    );

    const unsupported = UnsupprtedFields.map((field, i) => [
      field,
      `slice[${i}].data`,
    ]);

    test.each(unsupported)(
      "unsuported types %s should be falsy",
      (type, fieldText) => {
        expect(plugin.snippets({ type, fieldText })).toBeFalsy();
      }
    );

    test("repeatable slices", () => {
      const result = plugin.snippets({
        type: FieldType.Text,
        fieldText: "slices",
        useKey: false,
        isRepeatable: true,
      });

      expect(result).toEqual(
        '<div v-for="(item, i) in slice.items" :key="`slice-item-${i}`"><span>{{ slices }}</span></div>'
      );
    });
  });
});
