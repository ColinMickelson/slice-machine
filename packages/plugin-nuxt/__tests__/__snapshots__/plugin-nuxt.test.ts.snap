// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@slicemachine/plugin-nuxt #index it creates an index.js file 1`] = `
"// This file is generated by Slice Machine. Update with care!

export { default as Foo } from './Foo'
export { default as Bar } from './Bar'
export { default as Baz } from './Baz'
"
`;

exports[`@slicemachine/plugin-nuxt #slice 1`] = `
Object {
  "data": "<template>
  <section class=\\"section\\">
    <prismic-rich-text :field=\\"slice.primary.title\\" />
    <prismic-rich-text :field=\\"slice.primary.description\\" />
  </section>
</template>

<script>
export default {
  name: \\"MySlice\\",
  props: {
    slice: {
      type: Object,
      required: true,
      default() {
        return {}
      },
    },
  },
}
</script>

<style scoped>
.section {
  position: relative;
  background: #f7f7f7;
  color: #111;
  padding: 4em;
  text-align: center;
}
a {
  color: #111;
}
</style>
",
  "filename": "index.vue",
}
`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Boolean with useKey 1`] = `"<span>{{ slices[0].data ? 'true' : 'false' }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Color with useKey 1`] = `"<span :style=\\"\`color: \${slices[1].data}\`\\">Some Text</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet ContentRelationship with useKey 1`] = `"<prismic-link :field=\\"slices[2].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Date with useKey 1`] = `"<span>{{ slices[3].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Embed with useKey 1`] = `"<prismic-embed :field=\\"slices[4].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet GeoPoint with useKey 1`] = `"<span>{{ slices[5].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Image with useKey 1`] = `"<prismic-image :field=\\"slices[6].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Link with useKey 1`] = `"<prismic-link :field=\\"slices[7].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet LinkToMedia with useKey 1`] = `"<prismic-link :field=\\"slices[8].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Number with useKey 1`] = `"<span>{{ slices[9].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Select with useKey 1`] = `"<span>{{ slices[10].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet StructuredText with useKey 1`] = `"<prismic-rich-text :field=\\"slices[11].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Text with useKey 1`] = `"<span>{{ slices[12].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet Timestamp with useKey 1`] = `"<span>{{ slices[13].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets snippet UID with useKey 1`] = `"<span>{{ slices[14].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Boolean should be truthy 1`] = `"<span>{{ slices[0].data ? 'true' : 'false' }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Color should be truthy 1`] = `"<span :style=\\"\`color: \${slices[1].data}\`\\">Some Text</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type ContentRelationship should be truthy 1`] = `"<prismic-link :field=\\"slices[2].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Date should be truthy 1`] = `"<span>{{ slices[3].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Embed should be truthy 1`] = `"<prismic-embed :field=\\"slices[4].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type GeoPoint should be truthy 1`] = `"<span>{{ slices[5].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Image should be truthy 1`] = `"<prismic-image :field=\\"slices[6].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Link should be truthy 1`] = `"<prismic-link :field=\\"slices[7].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type LinkToMedia should be truthy 1`] = `"<prismic-link :field=\\"slices[8].data\\">My Link</prismic-link>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Number should be truthy 1`] = `"<span>{{ slices[9].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Select should be truthy 1`] = `"<span>{{ slices[10].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type StructuredText should be truthy 1`] = `"<prismic-rich-text :field=\\"slices[11].data\\" />"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Text should be truthy 1`] = `"<span>{{ slices[12].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type Timestamp should be truthy 1`] = `"<span>{{ slices[13].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #snippets supported type UID should be truthy 1`] = `"<span>{{ slices[14].data }}</span>"`;

exports[`@slicemachine/plugin-nuxt #story 1`] = `
Object {
  "data": "import MyComponent from '../slices/foo';
import SliceZone from 'vue-slicezone'

export default {
  title: 'FooBar'
}

export const default-slice = () => ({
  components: {
    MyComponent,
    SliceZone
  },
  methods: {
    resolve() {
      return MyComponent
    }
  },
  data() {
    return {
      mock: {\\"id\\":\\"default-slice\\",\\"name\\":\\"Default slice\\",\\"docURL\\":\\"...\\",\\"version\\":\\"sktwi1xtmkfgx8626\\",\\"description\\":\\"FooSlice\\",\\"primary\\":{\\"title\\":{\\"type\\":\\"StructuredText\\",\\"config\\":{\\"single\\":\\"heading1\\",\\"label\\":\\"Title\\",\\"placeholder\\":\\"This is where it all begins...\\"}},\\"description\\":{\\"type\\":\\"StructuredText\\",\\"config\\":{\\"single\\":\\"paragraph\\",\\"label\\":\\"Description\\",\\"placeholder\\":\\"A nice description of your feature\\"}},\\"\\":{\\"config\\":{\\"label\\":\\"\\",\\"constraint\\":{},\\"thumbnails\\":[]},\\"type\\":\\"Image\\"}},\\"items\\":{}}
    }
  },
  template: '<SliceZone :slices=\\"[mock]\\" :resolver=\\"resolve\\" />'
})
default-slice.storyName = 'Default slice'


",
  "filename": "index.stories.vue",
}
`;
